File: tools/sdks/ios/Sources/Thunder/auth/FlowExecutionClient.swift
import Foundation

/// Drives the Thunder Flow Execution API for app-native sign-in, sign-up, and recovery (spec §6.1–6.3).
final class FlowExecutionClient {
    private let httpClient: HTTPClient

    init(httpClient: HTTPClient) {
        self.httpClient = httpClient
    }

    func initiate(applicationId: String, flowType: FlowType) async throws -> EmbeddedFlowResponse {
        let body: [String: Any] = [
            "applicationId": applicationId,
            "flowType": flowType.rawValue,
            "verbose": true
        ]
        return try await httpClient.post(path: "/flow/execute", body: body, requiresAuth: false)
    }

    func submit(flowId: String, actionId: String, inputs: [String: String]) async throws -> EmbeddedFlowResponse {
        var body = submitBody(flowId: flowId, actionId: actionId)
        body["verbose"] = true
        if !inputs.isEmpty {
            body["inputs"] = inputs
        }
        return try await httpClient.post(path: "/flow/execute", body: body, requiresAuth: false)
    }

    func submitBody(flowId: String, actionId: String) -> [String: Any] {
        [
            "flowId": flowId,
            "action": actionId
        ]
    }
}

File: tools/sdks/thunderid-android/src/main/kotlin/io/thunder/android/auth/FlowExecutionClient.kt
package io.thunder.android.auth

import io.thunder.android.EmbeddedFlowResponse
import io.thunder.android.FlowType
import io.thunder.android.http.HttpClient

/**
 * Drives the Thunder Flow Execution API for app-native sign-in, sign-up, and recovery (spec §6.1–6.3).
 */
internal class FlowExecutionClient(private val httpClient: HttpClient) {

    suspend fun initiate(applicationId: String, flowType: FlowType): EmbeddedFlowResponse {
        val body = mapOf(
            "applicationId" to applicationId,
            "flowType" to flowType.value,
            "verbose" to true
        )
        return httpClient.post("/flow/execute", body, requiresAuth = false)
    }

    suspend fun submit(flowId: String, actionId: String, inputs: Map<String, String>): EmbeddedFlowResponse {
        val body = submitBody(flowId, actionId).toMutableMap()
        body["verbose"] = true
        if (inputs.isNotEmpty()) body["inputs"] = inputs
        return httpClient.post("/flow/execute", body, requiresAuth = false)
    }

    internal fun submitBody(flowId: String, actionId: String): Map<String, Any> = mapOf(
        "executionId" to flowId,
        "action" to actionId
    )
}
